File: frontend/src/entities/model-experiment/model/badges.ts
import type { StageName } from '@/entities/model-catalog/@x/model-experiment'

/** The fixed badge catalog a verdict may attach to each candidate. It is code-owned and its
 *  ids are stable, because a leaderboard tallies them across releases; the copy for each one
 *  lives in this slice's locale fragment. */
export type VerdictBadgeName =
  | 'fast'
  | 'natural'
  | 'on_brief'
  | 'structured'
  | 'accurate'
  | 'in_voice'
  | 'concise'
  | 'slow'
  | 'ai_like'
  | 'off_brief'
  | 'verbose'
  | 'inaccurate'
  | 'off_voice'
  | 'repetitive'
  | 'broken_format'
  | 'other'

/** The two groups the sheet presents, in the order it presents them. `other` belongs to
 *  neither and is rendered below both, because it carries a note rather than a judgement. */
export const POSITIVE_BADGES = [
  'fast',
  'natural',
  'on_brief',
  'structured',
  'accurate',
  'in_voice',
  'concise',
] as const satisfies readonly VerdictBadgeName[]

export const NEGATIVE_BADGES = [
  'slow',
  'ai_like',
  'off_brief',
  'verbose',
  'inaccurate',
  'off_voice',
  'repetitive',
  'broken_format',
] as const satisfies readonly VerdictBadgeName[]

/** The free note beside `other`, bounded the same way the server bounds it — in characters,
 *  so Korean prose is not cut to a third of what the field promises. */
export const BADGE_NOTE_MAX_LENGTH = 200

/** Whether a stage offers this badge. Only the voice pair is conditional: an observe
 *  comparison produces no prose, so neither judgement about voice can be made of it. */
export function badgeAppliesTo(badge: VerdictBadgeName, stage: StageName): boolean {
  if (badge === 'in_voice' || badge === 'off_voice') return stage === 'write' || stage === 'analyze'
  return true
}

/** Whether a badge reads as praise. Used for the tone a chip and a revealed badge carry. */
export function isPositiveBadge(badge: VerdictBadgeName): boolean {
  return (POSITIVE_BADGES as readonly VerdictBadgeName[]).includes(badge)
}

/** What one candidate was given at the verdict. Zero badges is as valid as ten. */
export interface CandidateBadges {
  candidateId: string
  badges: VerdictBadgeName[]
  otherNote: string
}

File: backend/internal/experiment/badges.go
package experiment

// The badge catalog is fixed and code-owned (MODEL-62): the operator does not edit it, and
// the ids are stable because a leaderboard tallies them across releases. Copy lives in the
// UI locale resources, never here.
type Badge string

const (
	// Positive.
	BadgeFast       Badge = "fast"
	BadgeNatural    Badge = "natural"
	BadgeOnBrief    Badge = "on_brief"
	BadgeStructured Badge = "structured"
	BadgeAccurate   Badge = "accurate"
	BadgeInVoice    Badge = "in_voice"
	BadgeConcise    Badge = "concise"
	// Negative.
	BadgeSlow         Badge = "slow"
	BadgeAILike       Badge = "ai_like"
	BadgeOffBrief     Badge = "off_brief"
	BadgeVerbose      Badge = "verbose"
	BadgeInaccurate   Badge = "inaccurate"
	BadgeOffVoice     Badge = "off_voice"
	BadgeRepetitive   Badge = "repetitive"
	BadgeBrokenFormat Badge = "broken_format"
	// Its own group, and the only one that carries a note.
	BadgeOther Badge = "other"
)

// BadgeNoteMaxLength bounds the free note beside `other`. Code-owned for the same reason the
// catalog is: the browser counts against it while typing and the server refuses past it.
const BadgeNoteMaxLength = 200

// PositiveBadges and NegativeBadges are the two groups the sheet presents, in the order it
// presents them. `other` belongs to neither and is rendered below both.
var (
	PositiveBadges = []Badge{BadgeFast, BadgeNatural, BadgeOnBrief, BadgeStructured, BadgeAccurate, BadgeInVoice, BadgeConcise}
	NegativeBadges = []Badge{BadgeSlow, BadgeAILike, BadgeOffBrief, BadgeVerbose, BadgeInaccurate, BadgeOffVoice, BadgeRepetitive, BadgeBrokenFormat}
)

// AppliesTo reports whether a stage offers this badge. Only the voice pair is conditional:
// an observe comparison produces no prose, so neither judgement about voice can be made of
// it (MODEL-62).
func (b Badge) AppliesTo(stage Stage) bool {
	if b == BadgeInVoice || b == BadgeOffVoice {
		return stage == StageWrite || stage == StageAnalyze
	}
	return true
}

func (b Badge) valid() bool {
	switch b {
	case BadgeFast, BadgeNatural, BadgeOnBrief, BadgeStructured, BadgeAccurate, BadgeInVoice, BadgeConcise,
		BadgeSlow, BadgeAILike, BadgeOffBrief, BadgeVerbose, BadgeInaccurate, BadgeOffVoice, BadgeRepetitive,
		BadgeBrokenFormat, BadgeOther:
		return true
	default:
		return false
	}
}

// CandidateBadges is what one candidate was given at the verdict.
type CandidateBadges struct {
	CandidateID string
	Badges      []Badge
	OtherNote   string
}

// NormalizeBadges checks the offered badges against the comparison they are being written
// onto and returns them deduplicated, in catalog order. Zero badges is a legal answer for
// every candidate — a badge is evidence offered, not a toll on the verdict (MODEL-61).
func NormalizeBadges(found Experiment, input []CandidateBadges) ([]CandidateBadges, error) {
	if len(input) == 0 {
		return nil, nil
	}
	known := map[string]bool{}
	for _, candidate := range found.Candidates {
		known[candidate.ID] = true
	}
	seenCandidate := map[string]bool{}
	out := make([]CandidateBadges, 0, len(input))
	for _, offered := range input {
		if !known[offered.CandidateID] || seenCandidate[offered.CandidateID] {
			return nil, ErrBadgesInvalid
		}
		seenCandidate[offered.CandidateID] = true
		chosen := map[Badge]bool{}
		for _, badge := range offered.Badges {
			if !badge.valid() || !badge.AppliesTo(found.Stage) {
				return nil, ErrBadgesInvalid
			}
			chosen[badge] = true
		}
		note := offered.OtherNote
		if note != "" && !chosen[BadgeOther] {
			return nil, ErrBadgesInvalid
		}
		if len([]rune(note)) > BadgeNoteMaxLength {
			return nil, ErrBadgesInvalid
		}
		ordered := make([]Badge, 0, len(chosen))
		for _, badge := range catalogOrder() {
			if chosen[badge] {
				ordered = append(ordered, badge)
			}
		}
		if len(ordered) == 0 && note == "" {
			continue
		}
		out = append(out, CandidateBadges{CandidateID: offered.CandidateID, Badges: ordered, OtherNote: note})
	}
	if len(out) == 0 {
		return nil, nil
	}
	return out, nil
}

func catalogOrder() []Badge {
	order := make([]Badge, 0, len(PositiveBadges)+len(NegativeBadges)+1)
	order = append(order, PositiveBadges...)
	order = append(order, NegativeBadges...)
	return append(order, BadgeOther)
}
